File: studio-idea-plugin/src/main/java/com/evolveum/midpoint/studio/ui/connector/generator/component/ElapsedTimerLabel.java
package com.evolveum.midpoint.studio.ui.connector.generator.component;

import com.intellij.ui.components.JBLabel;
import com.intellij.util.ui.UIUtil;

import javax.swing.*;
import javax.swing.Timer;

public class ElapsedTimerLabel extends JBLabel {

    private static final String FORMAT = "Elapsed time: %dm %ds";

    private long elapsedSeconds;
    private final Timer timer;

    public ElapsedTimerLabel(long initialSeconds) {
        super("", SwingConstants.CENTER);

        this.elapsedSeconds = initialSeconds;
        setForeground(UIUtil.getInactiveTextColor());
        updateText();

        timer = new Timer(1000, e -> {
            elapsedSeconds++;
            updateText();
        });
    }

    public void start() {
        timer.start();
    }

    public void stop() {
        timer.stop();
    }

    private void updateText() {

        setText(FORMAT.formatted(
                elapsedSeconds / 60,
                elapsedSeconds % 60
        ));
    }
}

File: studio-idea-plugin/src/main/kotlin/com/evolveum/midpoint/studio/ui/connector/generator/component/ElapsedTimerLabel.kt
/*
 * Copyright (C) 2010-2026 Evolveum and contributors
 *
 * Licensed under the EUPL-1.2 or later.
 */

package com.evolveum.midpoint.studio.ui.connector.generator.component

import com.intellij.ui.components.JBLabel
import com.intellij.util.ui.UIUtil
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.Timer

class ElapsedTimerLabel(private var elapsedSeconds: Long) : JBLabel("", CENTER) {
    private val timer: Timer

    init {
        setForeground(UIUtil.getInactiveTextColor())
        updateText()

        timer = Timer(1000, ActionListener { e: ActionEvent? ->
            elapsedSeconds++
            updateText()
        })
    }

    fun start() {
        timer.start()
    }

    fun stop() {
        timer.stop()
    }

    private fun updateText() {
        text = FORMAT.format(
            elapsedSeconds / 60,
            elapsedSeconds % 60
        )
    }

    companion object {
        private const val FORMAT = "Elapsed time: %dm %ds"
    }
}
